Natgrads

Daniel-Dodd opened the pull request:
<!--- Please provide a general summary of your changes in the title above -->

This PR seeks to add natural gradients to GPJax, as well as two new Gaussian variational family parameterisations.

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ x] Feature
- [ ] Code style update (formatting, renaming)
- [x ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

# Current state:

Code currently only works for the natural parameterisation case. Main thing (asides from the obvious simplification of the rough codebase and improvement of the API) is that we lack natural gradients for general parameterisations.

There are two notebooks associated with this PR:
- `natgrads.ipynb` has the case the variational family is chosen as the natural parameterisation - this should work but I have not tested it since I rebased with the master branch. 
- `Natural Gradient General case.ipynb` is provided to show a rough sketch of what the general case might look like, that will involve the user defining a bijection between their parameterisation and the natural parameterisation.

There are some unit tests provided for the new variational families, and for `natural_gradients.py` in its current state, as well as changes made in `parameters.py`. 

As a final note, it is likely that some of the functions defined in `natural_gradients.py` for stopping gradients might be better generalised and added to `abstractions.py` or `parameters.py`

Daniel-Dodd commented:
@thomaspinder it would be good to get your thoughts on the code so far (some of it rather rough). I believe the implementation works. The main issue is that the notebook needs writing before merging, I'll get on with this in the meantime. I'll need to test some benchmarks against `fit_batches` once #99 is merged, to see if performance is up to scratch.

thomaspinder commented on gpjax/natural_gradients.py:
```diff
@@ -0,0 +1,289 @@
+    natural_moments: dict, jitter: float = DEFAULT_JITTER
+) -> dict:
+    """
+    Converts natural parameters to expectation parameters.
```
Can we give the equations or at least a reference to the Salimbeni paper here?

thomaspinder commented on gpjax/natural_gradients.py:
```diff
@@ -0,0 +1,289 @@
+    return nat_grads_fn, hyper_grads_fn
+
+
+def fit_natgrads(
```
We need a unit test(s) for this fn.

thomaspinder reviewed (changes requested):
Started a review but will finish more tomorrow.

Daniel-Dodd commented:
@thomaspinder I have rebased the branch with master and would appreciate a review.

On my end, I need to:
- Update and add some tests
- Write the notebook

thomaspinder commented on docs/nbs/natgrads.ipynb:
```diff
@@ -0,0 +1,404 @@
+    "from gpjax.abstractions import progress_bar_scan\n",
+    "\n",
+    "#Set seed for reproducibility:\n",
+    "import tensorflow as tf\n",
+    "tf.random.set_seed(4)\n",
```
Redundant?

thomaspinder commented on gpjax/natural_gradients.py:
```diff
@@ -0,0 +1,288 @@
+
+
+def natural_to_expectation(
+    natural_moments: dict, jitter: float = DEFAULT_JITTER
```
Import `Dict` from typing in your imports and then use `Dict` as the type.

Daniel-Dodd replied:
Absolutely. This is something that is inconsistent across the package. Will address this in the next commit.

thomaspinder commented on gpjax/natural_gradients.py:
```diff
@@ -0,0 +1,288 @@
+    natural_moments: dict, jitter: float = DEFAULT_JITTER
+) -> dict:
+    """
+    Converts natural parameters to expectation parameters.
```
Can we add a little more mathematical context here on the operation that is being applied. Even if it's as simple as giving the equations. I know there in the inline comments, but this is what will be rendered in the docs.

Daniel-Dodd replied:
Good suggestion.

thomaspinder commented on gpjax/natural_gradients.py:
```diff
@@ -0,0 +1,288 @@
+        posterior: An instance of AbstractPosterior.
+        variational_family: An instance of AbstractVariationalFamily.
+    Returns:
+        Callable: A function that computes ELBO.
```
typo - the ELBO.

thomaspinder commented on gpjax/natural_gradients.py:
```diff
@@ -0,0 +1,288 @@
+    train_data: Dataset,
+) -> tp.Callable[[dict, Dataset], float]:
+    """
+    Construct evidence lower bound (ELBO) for varational Gaussian under the expectation parameterisation.
```
typo - variational

thomaspinder commented on gpjax/natural_gradients.py:
```diff
@@ -0,0 +1,288 @@
+    Returns:
+        Callable: A function that computes ELBO.
+    """
+    evg = ExpectationVariationalGaussian(
```
avoid variable names this short/ambiguous.

thomaspinder commented on gpjax/natural_gradients.py:
```diff
@@ -0,0 +1,288 @@
+    return svgp.elbo(train_data, identity_transformation, negative=True)
+
+
+def _stop_gradients_nonmoments(params: tp.Dict) -> tp.Dict:
+    """
+    Stops gradients for non-moment parameters.
+    Args:
+        params: A dictionary of parameters.
+    Returns:
+        tp.Dict: A dictionary of parameters with stopped gradients.
+    """
+    trainables = build_trainables_false(params)
+    moment_trainables = build_trainables_true(params["variational_family"]["moments"])
+    trainables["variational_family"]["moments"] = moment_trainables
+    params = trainable_params(params, trainables)
+    return params
+
+
+def _stop_gradients_moments(params: tp.Dict) -> tp.Dict:
+    """
+    Stops gradients for moment parameters.
+    Args:
+        params: A dictionary of parameters.
+    Returns:
+        tp.Dict: A dictionary of parameters with stopped gradients.
+    """
+    trainables = build_trainables_true(params)
+    moment_trainables = build_trainables_false(params["variational_family"]["moments"])
+    trainables["variational_family"]["moments"] = moment_trainables
+    params = trainable_params(params, trainables)
+    return params
```
could these two functions be combined and simplified. Something like `stop_gradients` with a boolean argument for natural parameters or moments? Seems excessive right now.

Daniel-Dodd replied:
Good suggestion. Will look into this.

thomaspinder commented on gpjax/natural_gradients.py:
```diff
@@ -0,0 +1,288 @@
+    # bijector = tp.Optional[dx.Bijector] = Identity, #bijector: A bijector to convert between the user chosen parameterisation and the natural parameters.
+) -> tp.Tuple[tp.Callable[[dict, Dataset], dict]]:
+    """
+    Computes natural gradients for variational Gaussian.
```
Be explicit about what the gradient is being computed with respect to i.e., the natural parameters.

thomaspinder commented on gpjax/natural_gradients.py:
```diff
@@ -0,0 +1,288 @@
+    stochastic_vi: StochasticVI,
+    train_data: Dataset,
+    transformations: dict,
+    # bijector = tp.Optional[dx.Bijector] = Identity, #bijector: A bijector to convert between the user chosen parameterisation and the natural parameters.
```
Can this be deleted?

thomaspinder commented on gpjax/natural_gradients.py:
```diff
@@ -0,0 +1,288 @@
+    return params
+
+
+def natural_gradients(
```
Is this fn. better suited to be in the abstractions file?

thomaspinder replied:
So are we moving this or not?

thomaspinder commented on gpjax/natural_gradients.py:
```diff
@@ -0,0 +1,288 @@
+    return nat_grads_fn, hyper_grads_fn
+
+
+def fit_natgrads(
```
As above - should this be moved into abstractions?

Daniel-Dodd replied:
I think that would be wise - was wondering the same.

thomaspinder commented on gpjax/natural_gradients.py:
```diff
@@ -0,0 +1,288 @@
+            nat_grad = dL_dexp
+            nat_grad["variational_family"]["moments"] = {
+                "natural_vector": dL_dexp["variational_family"]["moments"][
+                    "expectation_vector"
+                ],
+                "natural_matrix": dL_dexp["variational_family"]["moments"][
+                    "expectation_matrix"
+                ],
+            }
+
+            return value, nat_grad
+
+    else:
+        raise NotImplementedError
+
+    def hyper_grads_fn(params: dict, trainables: dict, batch: Dataset) -> dict:
+        """
+        Computes the hyperparameter gradients of the ELBO.
+        Args:
+            params: A dictionary of parameters.
+            trainables: A dictionary of trainables.
+            batch: A Dataset.
+        Returns:
+            dict: A dictionary of hyperparameter gradients.
+        """
+
+        def loss_fn(params: dict, batch: Dataset) -> f64["1"]:
+            # Determine hyperparameters that should be trained.
+            params = trainable_params(params, trainables)
+
+            # Stop gradients for the moment parameters.
+            params = _stop_gradients_moments(params)
+
+            return xi_elbo(params, batch)
+
+        value, dL_dhyper = value_and_grad(loss_fn)(params, batch)
+
+        return value, dL_dhyper
+
+    return nat_grads_fn, hyper_grads_fn
```
Fine for now, but this function is __very__ long (~100 lines). It'd be good if we could break it up in the future. If it's easy, then I'd suggest doing it within this PR. If not, please raise an issue to refactor this fn.

thomaspinder commented on gpjax/natural_gradients.py:
```diff
@@ -0,0 +1,288 @@
+    n_iters: tp.Optional[int] = 100,
+    log_rate: tp.Optional[int] = 10,
+) -> tp.Dict:
+
```
We should have a docstring here explaining why we optimise the variational distribution then the hyperparameters i.e., stale posteriors. Also happy for this information to be duplicated in the notebook.

thomaspinder commented on gpjax/parameters.py:
```diff
@@ -188,6 +187,20 @@ def transform(params: tp.Dict, transform_map: tp.Dict) -> tp.Dict:
+    # Copy dictionary structure
+    prior_container = deepcopy(params)
+
+    return jax.tree_map(lambda _: Identity.forward, prior_container)
```
jax.tree_util.tree_map

thomaspinder commented on gpjax/parameters.py:
```diff
@@ -278,6 +291,22 @@ def build_trainables(params: tp.Dict) -> tp.Dict:
+    # Copy dictionary structure
+    prior_container = deepcopy(params)
+    # Set all values to zero
+    prior_container = jax.tree_map(lambda _: False, prior_container)
```
tree_util

thomaspinder commented on gpjax/variational_families.py:
```diff
@@ -45,21 +45,27 @@ def predict(self, *args: Any, **kwargs: Any) -> dx.Distribution:
-    variational_mean: Optional[f64["M Q"]] = None
-    variational_root_covariance: Optional[f64["M M"]] = None
+    variational_mean: Optional[f64["N D"]] = None
+    variational_root_covariance: Optional[f64["N D"]] = None
```
Not `f64["N N"]`?

Daniel-Dodd replied:
Variational families are revised in d84ede9.

thomaspinder commented on gpjax/variational_families.py:
```diff
@@ -259,6 +267,282 @@ def predict_fn(test_inputs: f64["N D"]) -> dx.Distribution:
 
 
+@dataclass
+class NaturalVariationalGaussian(AbstractVariationalGaussian):
```
This object is wonderfully structured and coded.

thomaspinder reviewed (changes requested):
Several comments here, many of them suggestions, not hard requires. A general comment is that more information is needed in the docstrings - when the docs are built, this is what'll be rendered. I'm very happy for the maths within docstrings to then be moved into a notebook.

Daniel-Dodd commented:
@thomaspinder for the commits d84ede9, e0b60d4 and bd6e4aa, I have addressed many of your comments and suggestions.

Outstanding issues are to write unit tests, the notebook, and explanation of parameter optimisation order.

Daniel-Dodd commented:
Hi @thomaspinder, I believe I have addressed your comments so far. I expect all functions to be covered by unit tests now (but we'll see what CodeCov says). I would appreciate a review on the latest code changes - and I expect we can improve it further.

One key thing that I dislike, is that the `fit_natgrads` abstraction takes it fist argument as the variational inference strategy, while `fit` and `fit_batches` take in objectives. I'm not sure if you had any thoughts on this?


I am going to start work on writing the notebook today.

thomaspinder commented on gpjax/abstractions.py:
```diff
@@ -192,6 +194,80 @@ def step(carry, iter_num__and__key):
+    hyper_state = hyper_optim.init(params)
+    moment_state = moment_optim.init(params)
+
+    nat_grads_fn, hyper_grads_fn = natural_gradients(
+        stochastic_vi, train_data, transformations
+    )
```
These lines seem to be the only ones that require the `stochastic_vi` object. Can we simply move this outside of the training loop and then just make the function accept two objective_fn; one for the hyperparams and one for the natgrads?

thomaspinder commented on gpjax/abstractions.py:
```diff
@@ -192,6 +194,80 @@ def step(carry, iter_num__and__key):
+    )
+
+    # Initial natural gradient step to improve bound for hyperparameters:
+    batch = get_batch(train_data, batch_size, key)
+    loss_val, loss_gradient = nat_grads_fn(params, trainables, batch)
+    updates, moment_state = moment_optim.update(loss_gradient, moment_state, params)
+    params = optax.apply_updates(params, updates)
```
Do we want to run this for several iterations? We could perhaps add an optional argument similar to the number of warmup steps in a HMC sampler? Something like `natgrad_warmup_nits: int = 50`?

thomaspinder replied:
I'm fine with that.

thomaspinder commented on gpjax/parameters.py:
```diff
@@ -173,21 +172,35 @@ def inverse(bijector):
     )
 
 
+def build_identity(params: Dict) -> Dict:
+    """ "
+    Args:
+        params (Dict): The parameter set for which trainable statuses should be derived from.
+
+    Returns:
+        Dict: A dictionary of identity forward/backward bijectors. The dictionary is equal in structure to the input params dictionary.
+    """
+    # Copy dictionary structure
+    prior_container = deepcopy(params)
+
+    return jax.tree_util.tree_map(lambda _: Identity.forward, prior_container)
+
+
```
I don't get why this is needed?

Daniel-Dodd replied:
In an upcoming commit, I have changed `gpx.transform` to accept `transformation_map = None` to give an identity transformation.

Daniel-Dodd replied:
Resolved in https://github.com/thomaspinder/GPJax/commit/cb45fbfc3ffa449cb216c11250edb0fe4f20051b.

thomaspinder commented on gpjax/parameters.py:
```diff
@@ -278,12 +291,28 @@ def build_trainables(params: tp.Dict) -> tp.Dict:
+    prior_container = deepcopy(params)
+    # Set all values to zero
+    prior_container = jax.tree_util.tree_map(lambda _: False, prior_container)
+    return prior_container
```
I still don't like that this is two functions - it it feels like an argument would resolve it:
```python
def build_trainables_true(params: Dict, status: bool = False) -> Dict:
    """Construct a dictionary of trainable statuses for each parameter. By default, every parameter within the model is trainable.
    Args:
        params (Dict): The parameter set for which trainable statuses should be derived from.
        status (bool): Whether or not gradients should be taken with respect to this parameter.
    Returns:
        Dict: A dictionary of boolean trainability statuses. The dictionary is equal in structure to the input params dictionary.
    """
    # Copy dictionary structure
    prior_container = deepcopy(params)
    # Set all values to zero
    prior_container = jax.tree_util.tree_map(lambda _: status, prior_container)
    return prior_container

Daniel-Dodd replied:
Resolved in cb45fbf.

thomaspinder reviewed (changes requested):
A few small comments from me. When these are resolved and we're happy with the notebook, I think it's ready to merge.
